[Docs Rewrite] Create the new "Redux Fundamentals" tutorial

markerikson opened the pull request:
## PR Type :book: 

**Does this PR add a _new_ page, or update an _existing_ page?**

Adds the new **"Redux Fundamentals"** tutorial sequence

## Checklist

- [x] Is there an existing issue for this PR?
  - #3855
- [x] Have the files been linted and formatted?

## What docs page is being added or updated?

- **Section**: "Tutorials" > "Redux Fundamentals"
- **Page**: All of them!

## For Adding New Content

### What kind of content category is this page (tutorial, how-to, explanation, reference)?

Tutorial. 

### Who is the intended target audience?

Beginners who know nothing about Redux, yet.

#### What knowledge are we assuming they have?


- JS syntax: ES6, `async/await`, object spreads
- React: components, hooks
- Other: HTTP / AJAX, `fetch`

### What are the intended results or takeaways from reading this page?

Users should understand the fundamental concepts and principles of using Redux.

### What is the most critical info they should learn?

- Key Redux concepts and terms ("actions", "reducers", "state", "store", "dispatching")
- Core Redux data flow (dispatch -> reducer -> UI)
- Basic usage of Redux with React (`useSelector`, `useDispatch`, `<Provider>`)
- Standard Redux usage patterns (action creators, selectors, normalized state)
- To use Redux Toolkit instead of writing this code by hand


## Status

After having written the "Redux Essentials" tutorial, I'm now moving on to rewriting the existing "Basics/Advanced" tutorial sequence.

Per #3855 , the goals are:

- Drop all outdated references ("Flux", "containers", etc)
- Show simpler patterns (inline action types like `{type: "todos/addTodo"}` vs `const ADD_TODO = "ADD_TODO"`, single-file Redux logic)
- CodeSandbox examples
- Show React-Redux hooks as the default
- Improve explanation flow

and I've written up my plans and intended outline there as well.

Thus far, I've filled out the first two pages ("Overview", "Concepts and Data Flow") with content copy-pasted from the "Essentials" tutorial.  I've got the first half of page 3 ("State, Reducers, and Actions") thrown together, although I haven't worked on the actual example app at all yet.  I'll need to take time to do that shortly.

Happily, I've got a pretty good picture of how I want to tackle this app and explanation now.  I also will be able to reuse some of the explanations from the existing "Basics" tutorial, and I _intend_ to keep this shorter than the "Essentials" tutorial.  So, I'm hoping I'll be able to complete this one somewhat faster than the six months it took me to put the "Essentials" tutorial together :)

Shrugsy commented:
On the [Using middleware to enable async logic](https://deploy-preview-3894--redux-docs.netlify.app/tutorials/fundamentals/part-6-async-logic#using-middleware-to-enable-async-logic) section of part 6, there is the below middleware example: 

```js
const delayedActionMiddleware = storeAPI => next => action => {
  if (action.type === 'todos/todoAdded') {
    setTimeout(() => {
      // Delay this action by one second
      next(action)
    }, 1000)
  }

  return next(action)
}
```  

Currently the example still calls `next(action)` immediately, even for a `todos/todoAdded` action type, meaning it would be called once immediately, then again after 1 second instead of *only* delaying it by one second.

Shrugsy commented on docs/tutorials/fundamentals/part-7-standard-patterns.md:
```diff
@@ -0,0 +1,446 @@
+- Standard patterns used in real-world Redux apps, and why those patterns exist:
+  - Action creators for encapsulating action objects
+  - Memoized selectors for improving performance
+  - Normalizing state for manging collections of items
```
Minor typo here
```suggestion
  - Normalizing state for managing collections of items
```

juzhiyuan commented on docs/tutorials/fundamentals/part-3-state-actions-reducers.md:
```diff
@@ -332,10 +332,22 @@ We said earlier that **reducers must _always_ follow some special rules**:
+
+:::
+
+Any function that follows these rules is also known as a **"pure" function**, even if it's not specifically written as a reducer function.
```
`Any functions that follow those rules`

markerikson commented:
# First Draft Done!

It took me [6 months to write the "Essentials" tutorial](https://github.com/reduxjs/redux/pull/3740#issuecomment-650847906).  Happily, this one has only take about a month.  It's been a combination of having the experience from writing the "Essentials" tutorial, less stuff going on, and a few big days where I felt really inspired and made a lot of progress.

As a reminder, the PR preview is here:

**https://deploy-preview-3894--redux-docs.netlify.app/tutorials/fundamentals/part-1-overview**

the example app final source code / commits:

**https://github.com/reduxjs/redux-fundamentals-example-app/tree/tutorial-steps**

and a CodeSandbox with the final content:

**https://codesandbox.io/s/github/reduxjs/redux-fundamentals-example-app/tree/tutorial-steps**


## Thoughts

The "Essentials" tutorial ended up about 25K words.  The "Basics/Advanced" sequence is about 10K words.

Once again, I've hit about 25K words for the "Fundamentals" tutorial :)  That said, I did copy and paste some material from "Essentials", and there are a lot of code samples in here.

I think I've hit the goals that I was aiming for:

- Explained the what/why up front
- Clarified data flow
- Showed simpler patterns (inline action constants, single-file logic)
- Dropped outdated references like "Flux" and "container components"
- Showed why standard patterns like action creators exist
- Clarified how Redux actually works
- Showed how and why RTK is actually the right way to use Redux


## Requests for Review Focus

I'm specifically looking for review feedback in the following areas:

### Content Structure

- Does the overall layout and sequence of explanations flow logically?
- Do the page divisions and topics work?
- Specifically for the first page:
    - Should we keep the "What is Redux / State Management" sections on this page, to keep it a one-stop explanation, or should we move that content to somewhere under the "Introduction" category?
    - Right now the reader doesn't get into any hands-on coding until page 2.  Is there a way to reorganize things to get them involved sooner?
    - Do the Redux term explanations fit in here okay?
    - Right now the counter example walkthrough goes `store.js` -> `counterSlice.js` -> `Counter.js` -> `index.js`.  Should that be reordered, maybe starting with `index.js` first?
    
### Content Correctness

- Have I actually explained everything right?
- Are the explanations clear, especially for someone who's never used Redux before?
- Am I being too verbose?  Are there ways I can simplify explanations?
- Do the code samples match the explanations?
- I deliberately didn't show every diff, and hand-waved some changes.  Does this style of explanation work okay?
- Do the example repo / CodeSandbox clone and work correctly?
- Is the emphasis on immutability clear enough?

### Writing Approaches

- Does this section sufficiently follow the guidelines laid out in these "how to write good docs" guides?
  - [The "Tutorials" section in Divio's "The Documentation Guide"](https://documentation.divio.com/tutorials/)
  - [The style explanations in the Vue Docs Writing Guide](https://v3.vuejs.org/guide/contributing/writing-guide.html)

### Editing

- Any typos? Stuff I forgot to finish filling out?
- I know I need to finish filling in some links to Git branches and stuff, that's TBD


### Other

- How are the "What You'll Learn", "Prerequisites", and "What You Learned" boxes working out?
- Any other "Prerequisites" I should be listing for each page?
- Is using Docusaurus's code block line highlighting working out okay in terms of visibility and emphasis?
- Are the code block file titles sufficient?


## Remaining Work

I'd definitely like to get feedback and do some more polish before I publish it live.

The other big thing that needs to be done is adding redirect entries to point all external incoming link to the old tutorial pages over to relevant pages/sections of this tutorial instead, then remove the old pages entirely.

dceddia commented:
Congrats on getting this finished Mark! Huge accomplishment!

I played around with the example a bit and the color filters don't seem to stay checked. I think this line in `Footer.js` needs some braces:

```js
dispatch(colorFilterChanged(color, changeType))
```

versus:

```js
dispatch(colorFilterChanged({ color, changeType }))
```

I think the caveats in **How to Read This Tutorial** could dissuade people from reading, with the 2 bolded  sentences that roughly say "this tutorial is not the right way to do things".

Likewise the Info box that follows gives a "you're in the wrong place" vibe. I get the impression from forums and such that people will *absolutely* leap at the first mention of "how to use Redux, the real way", even if they have no background knowledge of it yet, so it might be good to tuck that away somewhere else 😄

Something along these lines might be good?

> This tutorial will teach you "how Redux works". Fair warning though - learning the concepts is different from putting them into practice in actual apps. 

> The initial code will be less concise than the way we suggest writing real app code, but writing it out long-hand is the best way to learn. Once you understand how everything fits together, we'll look at using Redux Toolkit to simplify things. Redux Toolkit is the recommended way to build production apps with Redux.

I like the State -> Actions -> Reducers -> Store -> UI progression in Part 1. The UI section would be a great place to call out that this is not React code, and in fact, Redux can with with any UI library.

Organization-wise, it was a little confusing that the second section (Redux Concepts and Data Flow) is actually called "Part 2". I think you could remove the "Part" from the titles and just refer to them as sections in the intro. Handy to have those numbers as I write this out, though...

The animation at the end of part 2 is great! I might even copy-paste it under the "What is Redux" section in Part 1.

I'm wondering if the more advanced Store stuff (middleware & store enhancers) would be better to have in a later section, maybe after "UI and React". I noticed the UI and React section didn't really do anything with middleware, but the Async Logic section uses it a bunch. Moving the initial explanation closer to where it's needed might be better for retention.

I like the code highlighting but I think the background color is a bit too light. The comments are hard to read over the lighter gray. The comments fail the AA contrast ratio in Chrome's devtools actually, even on the regular dark background. Making the comments a brighter `color: rgb(146 156 175)` and the highlighted lines a dark blue `#2d313e` seems to satisfy the contrast checker. With those changes the highlighted line background doesn't stand out very well from the regular background, so maybe some further tweaking is needed.

markerikson commented:
Thanks for the excellent feedback, Dave!  Quick responses:

- Yeah, I was changing the "color selected" implementation back and forth a couple times - it's likely I missed connecting the dots somewhere (although I coulda sworn it was working last time I checked)
- I like the rewording of the warning
- Not sure I follow the "Part" thing. Are you referring to the use of "Part N" in general?
- Agreed that we leave off the middleware stuff when we move from "Store" to "UI", but I decided it best fit into the "store configuration" section
- And yeah, we can totally change the CSS for code highlighting. Easy fix, and we should do that for all three of our sites.

JacopoDioBrando commented:
Hi,

I like this review better than the actual one but if I might, I'd suggest a couple of improvements or at least thoughts:
- I understand that Redux is not intended to be vertical onto just one UI library, but it's clearly most of the times associated and used with Redux. I'd consider using an example of passing data amongst two components (App and a child one?) and how passing data taking advantage of the whole Redux workflow could make the web app benefit
- a lot of people grew up with class components, "`mount`", connect, `mapToProps` and so on mantra...functional style is the way to do it, but maybe would make sense to explicit the transition here and why also Redux followed the same path?
- before there were action, creators and action types...how did we got here?
Unfortunately it's still plenty of recent tutorials that mention the old folder structure and way of thinking of coupling reducers, actions and so on...explaining the transition and compare the same example with the old and new way would help people to embrace the new way
- in the example default `store.dispatch` is used and people would might get confused that's the only way to do it; providing custom dispatching might help
- Maybe a small small example in TypeScript as well? :)

Dunno if I made myself 100% clear, but that's an awesome work of you, 'cause helps us noobs to follow the right patterns of improvement

markerikson commented:
@JacopoDioBrando appreciate the feedback!  I see where you're coming from, but I think the goals of this tutorial are a bit different than what you're describing.  Lemme respond to your thoughts:

- I assume one of those two "Redux" usages was meant to be "React".  In this case, we do briefly have an example of passing data from a parent component to a child component, right as we introduce `useSelector`: https://deploy-preview-3894--redux-docs.netlify.app/tutorials/fundamentals/part-5-ui-react#reading-state-from-the-store-with-useselector

```js
const renderedListItems = todos.map(todo => {
    return <TodoListItem key={todo.id} todo={todo} />
  })
```

- Per the whole "old Redux style" vs "new Redux style": I don't want this tutorial to do any comparisons between how we _used_ to do things, and how we _want_ people to do things.  For one thing, it's already 25000 words long :)  Also, there's already enough of an issue with the fact that most of the tutorial shows patterns that end up getting obsoleted by Redux Toolkit at the end.  If we start throwing in "well, we _used_ to have class components and `mapDispatch` and `/reducers` folders, BUT NONE OF THOSE EXIST ANY MORE!", I think it's going to make things a lot more confusing.
- Having said that, I can certainly see value in adding some kind of a "pattern migration" docs page that talks about some of that sort of thing, separate from this tutorial.
- I don't think I quite understood your point about `store.dispatch` and "custom dispatch".  Can you clarify that?
- Also definitely not showing any TS usage in this tutorial.  I tried that approach with the Redux Toolkit Advanced Tutorial, and got a number of folks pushing back saying it made things more confusing for them.  Given that the target audience for this tutorial includes "people who were just introduced to JS a few weeks ago in a bootcamp", adding TS would not help them.  We could definitely use a more detailed practical guide to writing Redux code with TS beyond the existing "Usage" page, though.

markerikson commented:
Pushed a bunch of updates. @dceddia , I've addressed all the points you listed except for the "middleware" ordering aspect, which I feel is fine as-is.  That includes updating the code block comment color to increase the contrast - it's now 9+ against the normal background, and 5-ish against the highlighted.

I've now gone through and updated all of the links that previously pointed to any of the "Basics/Advanced" Markdown files to point to relevant equivalents in the "Fundamentals" section instead.  I've also updated the `_redirects` file to handle external incoming links, and I've pulled the trigger on deleting the old Markdown files.  

I think this is basically ready to publish any time now.  I'll probably hold off until Monday morning just in case anyone else has feedback over the weekend.

Shrugsy commented:
Found that the dark mode colors are a bit messed up on this particular 'Info recap' section in 'part 8'

https://deploy-preview-3894--redux-docs.netlify.app/tutorials/fundamentals/part-8-modern-redux

![image](https://user-images.githubusercontent.com/22929669/97790358-3479ed80-1c1c-11eb-872b-3108c18b0c34.png)

markerikson commented:
Well, that's lovely.  I think we may have made a tweak on `master` at some point?  I need to merge down to fix the conflicts anyway

markerikson commented:
@Shrugsy yeah, looks like those blockquotes / "detailed" "blocks were just kinda broken in dark mode anyway.  I _think_ I've fixed them now.

phryneas commented:
As a general note: these codesandboxes have a lot of wasted screen space, especially on my ultrawide monitor.

Maybe we could have those flow out of the text to the left & right (or at least to the left)? I was thinking about [that full-bleed column technique by Josh Comeau](https://joshwcomeau.com/css/full-bleed/), but I have no idea how feasible that is in docusaurus.

![image](https://user-images.githubusercontent.com/4282439/97806583-9d815580-1c5c-11eb-90c4-a5d6bf36ec0e.png)

markerikson commented:
Yeah, that's more of a Docusaurus theme thing.  Worth noting, but really a separate topic :)

phryneas reviewed (commented):
Really liked it, and didn't really have a lot to citizize.

One point maybe: from reading the tutorial I would assume that it is just not possible to modify the redux state in a mutable way - only in the introduction of RTK you learn that there are things as "accidental mutations".
The whole concept of "mutating state is possible but it causes bugs" is not present anywhere else.

markerikson commented on docs/tutorials/fundamentals/part-1-overview.md:
```diff
@@ -0,0 +1,325 @@
+## What's Next?
+
+Now that you know what the basic pieces of a Redux app are, step ahead to [Part 2: Redux Concepts and Data Flow](./part-2-concepts-data-flow.md),
+where we'll look at how data flows through a Redux app in more detail.
```
Nothing we can do here as far as I know - Docusaurus is just pointing to whatever the prev/next page is sequentially in the entire ToC sidebar.

markerikson commented on docs/tutorials/fundamentals/part-1-overview.md:
```diff
@@ -0,0 +1,325 @@
+
+With that in mind, let's review what we've learned so far:
+
+:::tip
```
Updated to say "Summary" instead.

markerikson commented:
@phryneas addressed all your points locally.

I did cover examples of mutation in Part 3, but I'll add in a section in the "Store" page to show that you can (but shouldn't) do `getState().value = 123`.

markerikson commented:
Awright, I think this is pretty much good to go at this point.  Thanks to everyone who provided feedback - you've really helped polish this!
